### PR TITLE
Add objstore.List() recursive support

### LINKED ISSUE #3784: Object store clients: List() recursively
### Is your proposal related to a problem?

Cortex internally uses the Thanos objstore clients. We have an use case for which we need to list objects recursively, basically setting `delimiter=""` instead of the default `delimiter="/"`.

### Describe the solution you'd like

I'm proposing to extend `objstore.BucketReader` interface adding the following:

```
type IterOptions struct {
    // When true the iteration ignores the '/' delimiter.
    Recursive bool
}

type BucketReader interface {
    IterWithOptions(ctx context.Context, dir string, options IterOptions, f func(string) error) error
}
```

### Describe alternatives you've considered

- Add options directly to `Iter()` but would break signature add extra boilerplate to add to each `Iter()` call
- Add options are variadic at the end of `Iter()`, so `IterWithOptions(ctx context.Context, dir string, f func(string) error, options ...IterOption) error`: could work too, no big preference

### Additional context

Quick analysis of supported backends:
- S3
  - Minio already supports `Recursive` and we're currently hardcoding `Recursive: false`
- GCS
  - `storage.Query` allows to customise the `Delimiter` and we're currently hardcoding it to `/` (setting it to empty string returns the whole tree under the provided prefix, which is what we want with recursive = true)
- Azure
  - To list recursively we need to call `ListBlobsFlatSegment()` instead of `ListBlobsHierarchySegment()`
- Filesystem
  - We could switch from `ioutil.ReadDir()` to `filepath.Walk()` and recursively enter directories as soon as we find them (to keep lexicographically order guarantee)
- COS
  - If my chinese is good enough, according to the [doc](https://cloud.tencent.com/document/product/436/7734) it's just about passing an empty delimiter (supported by the SDK)
- OSS
  - Same here, according to [doc](https://www.alibabacloud.com/help/doc-detail/31965.htm) an empty delimiter will list recursively
- Swift
  - List objects with `swift.ObjectsOpts{ Delimiter: 0 }`

### PR DESCRIPTION
**Pull request by pracucci**
* [ ] I added CHANGELOG entry for this change.
* [x] Change is not relevant to the end user.

## Changes

As discussed in #3784, in this PR I'm proposing to add `IterOption` support to `objstore.Bucket.List()` supporting an option to enable recursive listing. This feature is not currently used by Thanos but would be used by Cortex (which depends on Thanos objstore clients).

Fixes #3784.

## Verification

Acceptance tests.

### CODE REVIEW

File: go.mod
Comment (pracucci) [inline go.mod:74]: To temporarily fix the Cortex compilation issue because of the changed `Iter()` signature. The little change is here:
https://github.com/thanos-io/thanos/compare/pracucci:add-list-recursively

I will open a follow up PR to remove this and upgrade Cortex once this PR will be merged (I will need to upgrade Thanos in Cortex, fix it in Cortex and then remove this override).

File: cmd/thanos/query_frontend.go
Comment (pracucci) [inline cmd/thanos/query_frontend.go:178]: Wasn't used by Thanos. Thanos use the tripperware below. Basically this code was doing nothing.

Comment (bwplotka) [reply cmd/thanos/query_frontend.go:178]: cc @yeya24 🤗

File: pkg/objstore/azure/azure.go
Comment (bwplotka) [inline pkg/objstore/azure/azure.go:119]: nit: (was that before): we could just put this in single line

File: pkg/objstore/objstore_test.go
Comment (bwplotka) [inline pkg/objstore/objstore_test.go:22]: Why this changed? 🤔

File: pkg/objstore/testing.go
Comment (bwplotka) [inline pkg/objstore/testing.go:172]: 👍🏽

File: go.mod
Comment (bwplotka) [inline go.mod:10]: Were you updating some deps too?

File: go.mod
Comment (bwplotka) [inline go.mod:81]: Is this needed? 🤗

**Review by bwplotka — APPROVED** [review]
Super nice, Mostly LGTM, some comments/questions though.

File: cmd/thanos/query_frontend.go
Comment (yeya24) [reply cmd/thanos/query_frontend.go:178]: Yes, it is not used so it is safe to remove it. Thanks @pracucci !

File: pkg/objstore/objstore_test.go
Comment (pracucci) [reply pkg/objstore/objstore_test.go:22]: Because I've changed the `AcceptanceTest` in `pkg/objstore/testing.go` adding more `Upload()` and `Iter()` calls.  This test just calls the `AcceptanceTest()` (see line above).

File: go.mod
Comment (pracucci) [reply go.mod:10]: I didn't. These has been picked up from the Cortex upgrade. If I remove the cortex `replace` directive and re-run `go mod vendor && go mod tidy` the `go.mod` rollbacks to the `master` version.

**Review by jtlisi — APPROVED** [review]
LGTM

This should allow us to efficiently transition the Cortex Alertmanager/Ruler clients to use the `objstore` implementation.

**Review by pstibrany — APPROVED** [review]
Nice work, lgtm.

**Comment by bwplotka** [comment]
The acceptance test @pracucci is actually failing after this PR. See https://github.com/thanos-io/thanos/pull/3841 https://app.circleci.com/pipelines/github/thanos-io/thanos/5124/workflows/e74a9507-75f8-4079-a9a3-bbe730866e3b/jobs/13814

I believe it's not has been run for your PR for some reasons. It might be because is it configured to only run for `write-access` user to avoid spam/malicious use of buckets. You have write access to repo, but maybe you need to log in to Circle CI and "follow" project?  Do you follow it here: https://app.circleci.com/projects/project-dashboard/github/thanos-io/ ?

Something is not working for GCS, so investigating.

File: go.mod
Comment (bwplotka) [reply go.mod:10]: 👍🏽

File: pkg/objstore/gcs/gcs.go
Comment (bwplotka) [inline pkg/objstore/gcs/gcs.go:98]: tiny bug, fixed here: https://github.com/thanos-io/thanos/pull/3841

Let's enable acceptance tests for you @pracucci 🤗

**Pull request merged**
